src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.NameGenerator.cs

```diff
@@ -0,0 +1,84 @@
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Words = System.Collections.Generic.IEnumerable<string>;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
+{
+    internal partial class DeclarationNameCompletionProvider
+    {
+        internal class NameGenerator
+        {
+            static readonly NamingStyle s_interfacePrefix = new NamingStyles.NamingStyle(Guid.NewGuid(),
+                prefix: "I", capitalizationScheme: Capitalization.PascalCase);
+
+            internal static ImmutableArray<Words> GetBaseNames(ITypeSymbol type)
+            {
+                var baseName = TryRemoveInterfacePrefix(type);
+                var breaks = StringBreaker.BreakIntoWordParts(baseName);
+                return GetInterleavedPatterns(breaks, baseName);
+            }
+
+            private static ImmutableArray<Words> GetInterleavedPatterns(StringBreaks breaks, string baseName)
+            {
+                var result = ImmutableArray.CreateBuilder<Words>();
+                result.Add(GetWords(0, breaks.Count, breaks, baseName));
+
+                for (int length = breaks.Count - 1; length > 0; length--)
+                {
+                    // going forward
+                    result.Add(GetLongestForwardSubsequence(length, breaks, baseName));
+
+                    // going backward
+                    result.Add(GetLongestBackwardSubsequence(length, breaks, baseName));
+                }
+                return result.ToImmutable();
+            }
+
+            private static Words GetLongestBackwardSubsequence(int length, StringBreaks breaks, string baseName)
+            {
+                var start = breaks.Count - length;
+                return GetWords(start, breaks.Count, breaks, baseName);
+            }
+
+            private static Words GetLongestForwardSubsequence(int length, StringBreaks breaks, string baseName)
+            {
+                var end = length;
+                return GetWords(0, end, breaks, baseName);
+            }
+
+            private static Words GetWords(int start, int end, StringBreaks breaks, string baseName)
+            {
+                var result = ImmutableArray.Create<string>();
+                for (; start < end; start++)
+                {
+                    var @break = breaks[start];
+                    result = result.Add(baseName.Substring(@break.Start, @break.Length));
+                }
+
+                return result;
+            }
+
+            private static string TryRemoveInterfacePrefix(ITypeSymbol type)
+            {
+                var name = type.Name;
+                if (type.TypeKind == TypeKind.Interface && name.Length > 1)
+                {
+                    string unused;
+                    if (s_interfacePrefix.IsNameCompliant(name, out unused))
+                    {
+                        return name.Substring(1);
+                    }
+                }
+
+                return type.CreateParameterName();
```

Comment:
It's the helper used by many feature to turn a type into a name like `v` or unwrap arrays. Not sure of a better name.